components/de_amp/browser/test/de_amp_browsertest.cc

```diff
@@ -0,0 +1,274 @@
+      body.c_str(), base_url.c_str(), port.c_str(), canonical_page.c_str()));
+  return http_response;
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, SimpleDeAmp) {
+  TogglePref(true);
+  ASSERT_TRUE(https_server_->Start());
+
+  // Go to any page
+  const GURL simple = https_server_->GetURL(kTestHost, kTestSimpleNonAmpPage);
+  ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), simple));
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), simple);
+
+  // Now go to an AMP page
+  https_server_.reset(new net::EmbeddedTestServer(
+      net::test_server::EmbeddedTestServer::TYPE_HTTPS));
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  const GURL landing_url = https_server_->GetURL(kTestHost, kTestCanonicalPage);
+  NavigateToURLAndWaitForRedirects(original_url, landing_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, NonHttpScheme) {
+  TogglePref(true);
+  const std::string nonHttpSchemeBody =
+      R"(
+        <html amp>\
+      <head>\
+      <link rel='canonical' 
+      href='brave://settings'>\
+      </head></html>
+      )";
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, nonHttpSchemeBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, CanonicalLinkSameAsAmpPage) {
+  TogglePref(true);
+  https_server_->RegisterRequestHandler(
+      base::BindRepeating(HandleRequest, kTestHost, kTestAmpPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, PrefOff) {
+  TogglePref(false);
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  // Doesn't get De-AMPed
+  NavigateToURLAndWaitForRedirects(original_url, original_url);
+}
+
+IN_PROC_BROWSER_TEST_F(DeAmpBrowserTest, RestorePage) {
+  TogglePref(true);
+  https_server_->RegisterRequestHandler(base::BindRepeating(
+      HandleRequest, kTestHost, kTestCanonicalPage, kTestBody));
+  ASSERT_TRUE(https_server_->Start());
+  const GURL original_url = https_server_->GetURL(kTestHost, kTestAmpPage);
+  const GURL landing_url = https_server_->GetURL(kTestHost, kTestCanonicalPage);
+  NavigateToURLAndWaitForRedirects(original_url, landing_url);
+
+  // Close & restore tab
+  browser()->tab_strip_model()->CloseSelectedTabs();
+  Profile* profile = browser()->profile();
```

Comment:
we always do by default in a browser test and it would have one tab by default

Reply:
@bridiver 
We call `CloseSelectedTabs()` at the previous line, don't we?

Reply:
Not sure I follow, why is that an issue?

Reply:
If we close the browser, them browser() is invalid pointer. Also I'm not sure that it's a stable state without KeepAlive objects.

Reply:
@atuchin-m oh, I see what you're saying. Looks like we also have a potential race condition here because we should have something like 
```
content::WebContents* tab_to_close =
      browser()->tab_strip_model()->GetActiveWebContents();
  content::WebContentsDestroyedWatcher destroyed_watcher(tab_to_close);
  browser()->tab_strip_model()->CloseSelectedTabs();
  destroyed_watcher.Wait();
```

Reply:
although I also see this so I'd have to dig more to see how this is supposed to work
```
// Close the current tab.
    browser()->tab_strip_model()->CloseSelectedTabs();

    GURL url3("chrome://chrome-urls");
    ui_test_utils::NavigateToURLWithDisposition(
        browser(), url3, WindowOpenDisposition::CURRENT_TAB,
        ui_test_utils::BROWSER_TEST_WAIT_FOR_LOAD_STOP);
```